thelounge: fix build

winterqt opened the pull request:
###### Description of changes

Upstream switched to using TypeScript in v4.4.0, which broke the patch. This fixes that issue by migrating to building The Lounge from source, instead of having to patch the minified JavaScript.

Closes #233630.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

winterqt committed:
thelounge: fix build

Upstream switched to using TypeScript in v4.4.0, which broke the patch.
This fixes that issue by migrating to building The Lounge from source,
instead of having to patch the minified JavaScript.

lilyinstarlight commented on pkgs/applications/networking/irc/thelounge/default.nix:
```diff
@@ -0,0 +1,65 @@
+    fixup_yarn_lock yarn.lock
+    yarn config --offline set yarn-offline-mirror ${finalAttrs.offlineCache}
+    yarn install --offline --frozen-lockfile --ignore-platform --ignore-scripts --no-progress --non-interactive
+    patchShebangs node_modules
```
buildYarnPackage hooks when

winterqt replied:
Something something #210814

lilyinstarlight reviewed (commented):
Assuming ofborg shows all green, this looks good to me (really want those yarn hooks though and my PR to add a fixup-yarn-lock command independent of yarn2nix...)

RaitoBezarius commented:
I'm going to be a guinea pig for that, pushing in production.

RaitoBezarius commented:
Production is running with it, without any hiccups. :)

RaitoBezarius commented on pkgs/applications/networking/irc/thelounge/default.nix:
```diff
@@ -0,0 +1,65 @@
+    description = "A glossy Matrix collaboration client for the web";
+    homepage = "https://thelounge.chat";
+    changelog = "https://github.com/thelounge/thelounge/releases/tag/v${finalAttrs.version}";
+    maintainers = with maintainers; [ winter ];
```
you can add me too as a maintainer, no problem

winterqt replied:
Sure, thanks.

RaitoBezarius reviewed (approved):
LGTM!

RaitoBezarius commented on pkgs/applications/networking/irc/thelounge/default.nix:
```diff
@@ -0,0 +1,65 @@
+  passthru.tests = nixosTests.thelounge;
+
+  meta = with lib; {
+    description = "A glossy Matrix collaboration client for the web";
```
yes yes it is :DDDDDDDDDDDDd

winterqt commented:
Waiting on OfBorg.

winterqt commented:
Actually, let me try something else... I want to reduce this closure size.

winterqt committed:
npmHooks.npmInstallHook: allow disabling `npm prune` invocation

In some odd scenarios, `npm prune` either fails, or hangs. I have no idea
what could possibly be wrong at the moment, but let's provide an escape
hatch for packages that can still use the rest of the install hook's
functionality.

winterqt commented:
Before this push, it was just about 400 MiB, now it's 81 MiB.

RaitoBezarius reviewed (approved):
LGTM

SuperSandro2000 commented on pkgs/development/node-packages/aliases.nix:
```diff
@@ -44,4 +44,5 @@ mapAliases {
   "@nestjs/cli" = pkgs.nest-cli; # Added 2023-05-06
   manta = pkgs.node-manta; # Added 2023-05-06
   triton = pkgs.triton; # Added 2023-05-06
+  thelounge = pkgs.thelounge; # Added 2023-05-22
```
```suggestion
  thelounge = pkgs.thelounge; # Added 2023-05-22
  triton = pkgs.triton; # Added 2023-05-06
```
nit

winterqt replied:
Thanks.

SuperSandro2000 commented on pkgs/applications/networking/irc/thelounge/default.nix:
```diff
@@ -0,0 +1,71 @@
+  buildPhase = ''
+    runHook preBuild
+
+    NODE_ENV=production yarn build
```
we should also move the env to some hook but out of scope

winterqt replied:
Agreed. However, this isn't the time for it due to the whole tooling shuffling/etc.

RaitoBezarius commented on pkgs/build-support/node/build-npm-package/hooks/npm-install-hook.sh:
```diff
@@ -27,7 +27,10 @@ npmInstallHook() {
 
     if [ ! -d "$nodeModulesPath" ]; then
-        npm prune --omit dev --no-save $npmInstallFlags "${npmInstallFlagsArray[@]}" $npmFlags "${npmFlagsArray[@]}"
+        if [ -z "${dontNpmPrune-}" ]; then
```
This will be taken care in further PRs.

winterqt replied:
Any reason why we shouldn't just do it here?

winterqt replied:
Actually, we don't have any of the other `dont*` flags here, so let's not.

lilyinstarlight reviewed (approved):
This looks good to me too. I'll second Sandro's minor lexicographical ordering nit, but the `dontNpmPrune` install hook change looks fine enough for me

(I wish it were reasonable to separate prune/install into separate hooks entirely, but it really sorta has to happen in the middle there, so this is acceptable for now)

winterqt commented:
Should be good to go now. Waiting on OfBorg.

winterqt commented:
I need more details, please -- what system are you running it on? This seemingly can't be reproduced in CI, or on Ryan's end.

shyim commented:
I have the same aarch64-linux my config: https://github.com/shyim/nixcfg/blob/main/systems/shea/services/thelounge.nix

Tom-Hubrecht commented:
> I need more details, please -- what system are you running it on? This seemingly can't be reproduced in CI, or on Ryan's end.

I am running thelounge on an x86_64 server, the config is at https://git.hubrecht.ovh/hubrecht/nixos/src/branch/master/machines/sd-02/thelounge.nix

The issue seems to be that old messages are no longer shown, but new messages still are (I am next to Ryan's guinea pig who confirmed the issue)

RaitoBezarius commented:
> I need more details, please -- what system are you running it on? This seemingly can't be reproduced in CI, or on Ryan's end.

I confirm that my users doesn't see their messages, but the logs are on disk and the SQLite DB is intact.

winterqt commented:
Let me look into this, please give me a bit.

RaitoBezarius commented:
I do have this:
```
May 23 16:15:00 amadeus thelounge[108838]: 2023-05-23 14:15:00 [ERROR] Unable to load sqlite3 module. See https://github.com/mapbox/node-sqlite3/wiki/Binaries
May 23 16:15:00 amadeus thelounge[108838]: 2023-05-23 14:15:00 [INFO] The Lounge v4.4.0 (Node.js 18.16.0 on linux x64)
May 23 16:15:00 amadeus thelounge[108838]: 2023-05-23 14:15:00 [INFO] Configuration file: /var/lib/thelounge/config.js
May 23 16:15:01 amadeus thelounge[108838]: 2023-05-23 14:15:01 [INFO] Available at http://127.0.0.1:9000/ in private mode
```

in my systemd logs, FWIW.

@winterqt Thank you for looking into it (and I have no worries, I have backups and all.) :)

It's annoying that upstream didn't make the error fatal… :/
